include/perfetto/protozero/proto_decoder.h

```diff
@@ -267,6 +272,61 @@ class PackedRepeatedFieldIterator {
   bool* const parse_error_;
 };
 
+// A dispatching wrapper that can handle both packed and non-packed repeated
+// fields. It holds either a RepeatedFieldIterator or
+// PackedRepeatedFieldIterator and forwards calls to the appropriate one using
+// std::variant and base utilities.
+template <proto_utils::ProtoWireType wire_type, typename CppType>
+class UnifiedRepeatedFieldIterator {
+ public:
+  using RegularIterator = RepeatedFieldIterator<CppType>;
+  using PackedIterator = PackedRepeatedFieldIterator<wire_type, CppType>;
+  using VariantType = std::variant<RegularIterator, PackedIterator>;
+
+  // Constructor taking a regular iterator
+  explicit UnifiedRepeatedFieldIterator(RegularIterator iter)
+      : iter_(std::move(iter)) {}
+
+  // Constructor taking a packed iterator
+  explicit UnifiedRepeatedFieldIterator(PackedIterator iter)
+      : iter_(std::move(iter)) {}
+
+  // Constructs an invalid iterator.
+  UnifiedRepeatedFieldIterator() : iter_(RegularIterator()) {}
+
+  explicit operator bool() const {
+    if (std::holds_alternative<RegularIterator>(iter_)) {
```

Comment:
note that you could avoid one branch here, as internally get_if already does a branch (because returns null if you hold it wrong),
so this could be like:

if (auto* r = std::get_if<RegularIterator>(&iter_))
  return static_cast<bool>(*r);
return static_cast<bool>(*std::get_if<PackedIterator>(&iter_));


same below everywhere

or technically you could even just std::visit to keep everything more compact

return std::visit([](auto& it){ return static_cast<bool>(it); }, iter_);

I checked with godbolt with a minified version and emits very miminmal code with only one branch: https://godbolt.org/z/5jzsarKPz